basic_sender.go

```diff
@@ -0,0 +1,362 @@
+	return opts.logID
+}
+
+// NewLogger returns a grip Sender backed by Cedar Buildlogger with level
+// information set.
+func NewLogger(ctx context.Context, name string, l send.LevelInfo, opts *LoggerOptions) (send.Sender, error) {
+	b, err := MakeLogger(ctx, name, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem making new logger")
+	}
+
+	if err := b.SetLevel(l); err != nil {
+		return nil, errors.Wrap(err, "problem setting grip level")
+	}
+
+	return b, nil
+}
+
+// MakeLogger returns a grip Sender backed by Cedar Buildlogger.
+func MakeLogger(ctx context.Context, name string, opts *LoggerOptions) (send.Sender, error) {
+	ts := time.Now()
+
+	if err := opts.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid cedar buildlogger options")
+	}
+
+	var conn *grpc.ClientConn
+	var err error
+	if opts.ClientConn == nil {
+		rpcOpts := []grpc.DialOption{
+			grpc.WithUnaryInterceptor(aviation.MakeRetryUnaryClientInterceptor(10)),
+			grpc.WithStreamInterceptor(aviation.MakeRetryStreamClientInterceptor(10)),
+		}
+		if opts.Insecure {
+			rpcOpts = append(rpcOpts, grpc.WithInsecure())
+		} else {
+			var tlsConf *tls.Config
+			tlsConf, err = aviation.GetClientTLSConfig(opts.CAFile, opts.CertFile, opts.KeyFile)
+			if err != nil {
+				return nil, errors.Wrap(err, "problem getting client TLS config")
+			}
+
+			rpcOpts = append(rpcOpts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConf)))
+		}
+
+		conn, err := grpc.DialContext(ctx, opts.RPCAddress, rpcOpts...)
+		if err != nil {
+			return nil, errors.Wrap(err, "problem dialing rpc server")
+		}
+		opts.ClientConn = conn
+	}
+
+	b := &buildlogger{
+		ctx:    ctx,
+		opts:   opts,
+		conn:   conn,
+		client: internal.NewBuildloggerClient(opts.ClientConn),
+		buffer: []*internal.LogLine{},
+		Base:   send.NewBase(name),
+	}
+
+	if err := b.SetErrorHandler(send.ErrorHandlerFromSender(b.opts.Local)); err != nil {
+		return nil, errors.Wrap(err, "problem setting default error handler")
+	}
+
+	if err := b.createNewLog(ts); err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// Send sends the given message with a timestamp created when the function is
+// called to the Cedar Buildlogger backend. This function buffers the messages
+// until the maximum allowed buffer size is reached, at which point the
+// messages in the buffer are sent to the Buildlogger server via RPC. Send is
+// thread safe.
+func (b *buildlogger) Send(m message.Composer) {
+	if !b.Level().ShouldLog(m) {
+		return
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	ts := time.Now()
+	if b.closed {
+		b.opts.Local.Send(message.NewErrorMessage(level.Error, errors.New("cannot call Send on a closed Buildlogger Sender")))
+		return
+	}
+
+	_, ok := m.(*message.GroupComposer)
+	var lines []string
+	if b.opts.NewLineCheckOff && !ok {
+		lines = []string{m.String()}
+	} else {
+		lines = strings.Split(m.String(), "\n")
+	}
+
+	for _, line := range lines {
+		if line == "" {
+			continue
+		}
```

Comment:
I think we have to remove this.

Reply:
I don't think so, new lines are preserved (they are added back on in cedar) and removing this may add extra new lines in cases where there was a newline at the end of an incoming line.